src/hooks/useDragAndDropSubThought.ts

```diff
@@ -101,66 +101,152 @@ const drop = (props: DroppableSubthoughts, monitor: DropTargetMonitor) => {
+    const thoughtTo = getThoughtById(state, head(rootedParentOf(state, pathTo)))
+    const thoughtFrom = getThoughtById(state, head(thoughtsFrom))
+    const parentIdFrom = head(rootedParentOf(state, thoughtsFrom))
+    const parentIdTo = head(rootedParentOf(state, pathTo))
+    const sameContext = parentIdFrom === parentIdTo
+
+    // cannot drop on itself
+    if (!thoughtFrom || !thoughtTo || equalPath(thoughtsFrom, thoughtsTo)) return { isValid: false }
+
+    // cannot move root or em context or target is divider
+    if (isDivider(thoughtTo?.value) || (isRootOrEM && !sameContext)) {
+      return {
+        isValid: false,
+        errorMessage: `Cannot move the ${isEM(thoughtsFrom) ? 'em' : 'home'} context to another context.`,
+        errorType: 'error',
+      }
+    }
+
+    return { isValid: true }
   }
 
-  haptics.medium()
+  const state = store.getState()
+
+  // If any drop is invalid, abort the drop early
+  if (
+    draggedItems.some(item => {
+      const { isValid, errorMessage, errorType } = validateDraggedItem(state, item, props.simplePath)
 
-  store.dispatch(
-    moveThought({
-      oldPath: thoughtsFrom,
-      newPath: pathTo,
-      newRank: (dropTop ? getPrevRank : getNextRank)(state, thoughtTo.id),
-    }),
+      if (!isValid && errorMessage) {
+        if (errorType === 'warning') {
+          console.warn(errorMessage, { item })
+        } else if (errorType === 'error') {
+          store.dispatch(error({ value: errorMessage }))
+        }
+      }
+
+      return !isValid
+    })
   )
+    return
+
+  /** Check if the context has changed for all dragged items. */
+  const hasContextChanged = draggedItems.every(item => {
+    const pathTo = getPathTo(state, item.path)
+    return head(rootedParentOf(state, item.path)) !== head(rootedParentOf(state, pathTo))
+  })
+
+  store.dispatch((dispatch, getState) => {
+    /** Returns the parent ID and drop top flag for a given path. */
+    const getDropParams = (pathTo: Path) => {
+      const parentPath = rootedParentOf(getState(), pathTo)
+      const parentId = head(parentPath)
+      const isExpanded = isPathExpanded(getState(), parentPath)
+      const dropTop = !isExpanded && attributeEquals(getState(), parentId, '=drop', 'top')
 
-  // alert user of move to another context
-  if (!sameContext) {
-    // wait until after MultiGesture has cleared the error so this alert does no get cleared
-    setTimeout(() => {
-      const alertFrom = '"' + ellipsize(thoughtFrom.value) + '"'
-      const alertTo = parentIdTo === HOME_TOKEN ? 'home' : '"' + ellipsize(thoughtTo.value) + '"'
-      const inContext = props.showContexts
-        ? ` in the context of ${ellipsize(headValue(state, props.simplePath ?? props.path) ?? 'MISSING_CONTEXT')}`
-        : ''
-
-      store.dispatch(
-        alert(`${alertFrom} moved to${dropTop ? ' top of' : ''} ${alertTo}${inContext}.`, {
-          clearDelay: 5000,
+      return { parentId, dropTop }
```

Comment:
Try to avoid functions that return groups of values as an object. It's an ugly pattern. Remember that functions are meaningful units of work, not arbitrary bundles of functionality. There is no singular concept that corresponds to `{ parentId, dropTop }`.

I think it would be okay to return `dropTop` directly. `parentId` is easily accessed from the `Path` with `head(rootedParentOf(state, pathTo))` and does not need to be abstracted.